libqpdf/QPDF.cc

```diff
@@ -2376,19 +2376,26 @@ QPDF::getCompressibleObjGens()
     QPDFObjectHandle encryption_dict = m->trailer.getKey("/Encrypt");
     QPDFObjGen encryption_dict_og = encryption_dict.getObjGen();
 
-    QPDFObjGen::set visited;
-    std::list<QPDFObjectHandle> queue;
-    queue.push_front(m->trailer);
+    const size_t max_obj = getObjectCount();
+    std::vector<bool> visited(max_obj, false);
+    std::vector<QPDFObjectHandle> queue;
+    queue.reserve(512);
+    queue.push_back(m->trailer);
     std::vector<QPDFObjGen> result;
     while (!queue.empty()) {
-        QPDFObjectHandle obj = queue.front();
-        queue.pop_front();
+        auto obj = queue.back();
+        queue.pop_back();
         if (obj.isIndirect()) {
             QPDFObjGen og = obj.getObjGen();
-            if (!visited.add(og)) {
+            const size_t id = toS(og.getObj() - 1);
+            if (id >= max_obj)
+                throw std::runtime_error(
```

Comment:
It is not correct for this to be a `std::runtime_error`. If this can only happen because of a programming error, it should be a `std::logic_error`. If it could happen because of something about the input file, which I doubt, unless `getObjectCount` is broken or there is concurrent access, which is not allowed, then it should be handled more gracefully. By the time this is called, we're already in QPDFWriter.

Reply:
Agree this should be a logic error.